src/event.rs

```diff
@@ -250,34 +253,42 @@ impl<N: Notify> Processor<N> {
                 }
             }
 
-            match window.wait_events().next() {
+            let event = if self.wait_for_event {
+                window.wait_events().next()
+            } else {
+                None
+            };
+
+            terminal = term.lock();
+
+            context = ActionContext {
+                terminal: &mut terminal,
+                notifier: &mut self.notifier,
+                selection: &mut self.selection,
+                mouse: &mut self.mouse,
+                size_info: &self.size_info,
+            };
+
+            processor = input::Processor {
+                ctx: context,
+                key_bindings: &self.key_bindings[..],
+                mouse_bindings: &self.mouse_bindings[..]
+            };
+
+            match event {
                 Some(event) => {
-                    terminal = term.lock();
-                    context = ActionContext {
-                        terminal: &mut terminal,
-                        notifier: &mut self.notifier,
-                        selection: &mut self.selection,
-                        mouse: &mut self.mouse,
-                        size_info: &self.size_info,
-                    };
-
-                    processor = input::Processor {
-                        ctx: context,
-                        key_bindings: &self.key_bindings[..],
-                        mouse_bindings: &self.mouse_bindings[..]
-                    };
-
                     process!(event);
                 },
-                // Glutin guarantees the WaitEventsIterator never returns None.
-                None => unreachable!(),
+                None => ()
             }
 
             for event in window.poll_events() {
                 process!(event);
             }
         }
 
+        self.wait_for_event = !terminal.dirty;
```

Comment:
I'm struggling to make this change—I can't compute `wait_for_event` from within the `process_events`, because it depends on `terminal` which isn't initialized until after calling `window.wait_events().next()` (itself gated by `wait_for_event`).

So I've also looked at passing in `wait_for_event` as an argument to `process_events` in `src/main.rs`. Each time in the loop, I could update `wait_for_event` after `draw` based on `completed`, but then I hit a compile error because I think `draw` takes ownership of the `terminal`? Anyway, before I go further, does it sound like I am on the right track?